### DISCUSSION TITLE
[Feature Request]: Sharing Sites On Your Local Network

### DISCUSSION THREAD

**Discussion by slavarazum** [beyondcode/herd-community#357] [Ideas]
### Feature Description

# Feature Description

The `herd share` suites are for public sharing and do not support linking to local resources.
The development build with `npm run dev` is not available through the expose tunnel.

From the original Valet feature: https://laravel.com/docs/master/valet#sharing-sites-on-your-local-network

The feature allows you to share sites on your local network, making it easier to set up development environments with real-time access on mobile or other network-accessible devices.

## Is this feature valuable for other users as well and why?

Assets building with hot reload will be available for all local accessible devices.

### Is this feature valuable for other users as well and why

Development with a real device is a special pleasure 📱🌴

**Comment by pLavrenov** [comment]
up

**Comment by belameiers** [comment]
That would be awesome. I have the same problem and always need to switch to MAMP for local testing. The Expose Function is as well not an option for me…

**Comment by Gnative** [comment]
This is the one thing preventing me from uninstalling Valet and using only Herd.

I work with mobile development and need to access my Lavarel application API endpoints on my local network. With Valet I can access this via my local network IP address I.E http://192.168.8.8/contours.test/api/v1.

I haven't been able to find a way to do this with Herd yet unless i've missed something in the documentation.

**Comment by Gnative** [comment]
Ok just realised I can open up local network IP address the same as Valet by editing 

`~/Library/Application Support/Herd/config/nginx/herd.conf`

changing 

`listen 127.0.0.1:80 default_server;`

to 

`listen 80 default_server;`



```
server {
    listen 80 default_server;
    #listen VALET_LOOPBACK:80; # valet loopback
    root /;
    charset utf-8;
    client_max_body_size 128M;

    location /41c270e4-5535-4daa-b23e-c269744c2f45/ {
        internal;
        alias /;
        try_files $uri $uri/;
    }

    location / {
        rewrite ^ "/Applications/Herd.app/Contents/Resources/valet/server.php" last;
    }

    location = /favicon.ico { access_log off; log_not_found off; }
    location = /robots.txt  { access_log off; log_not_found off; }

    access_log off;
    error_log "/Users/Gnative/Library/Application Support/Herd/Log/nginx-error.log";

    error_page 404 "/Applications/Herd.app/Contents/Resources/valet/server.php";

    location ~ [^/]\.php(/|$) {
        fastcgi_split_path_info ^(.+\.php)(/.+)$;
        fastcgi_pass $herd_sock;
        fastcgi_index "/Applications/Herd.app/Contents/Resources/valet/server.php";
        fastcgi_param QUERY_STRING  $query_string;
        fastcgi_param REQUEST_METHOD  $request_method;
        fastcgi_param CONTENT_TYPE  $content_type;
        fastcgi_param CONTENT_LENGTH  $content_length;
        fastcgi_param SCRIPT_FILENAME  $request_filename;
        fastcgi_param SCRIPT_NAME  $fastcgi_script_name;
        fastcgi_param REQUEST_URI  $request_uri;
        fastcgi_param DOCUMENT_URI  $document_uri;
        fastcgi_param DOCUMENT_ROOT  $document_root;
        fastcgi_param SERVER_PROTOCOL  $server_protocol;
        fastcgi_param GATEWAY_INTERFACE CGI/1.1;
        fastcgi_param SERVER_SOFTWARE  nginx/$nginx_version;
        fastcgi_param REMOTE_ADDR  $remote_addr;
        fastcgi_param REMOTE_PORT  $remote_port;
        fastcgi_param SERVER_ADDR  $server_addr;
        fastcgi_param SERVER_PORT  $server_port;
        fastcgi_param SERVER_NAME  $server_name;
        fastcgi_param HTTPS   $https if_not_empty;
        fastcgi_param HERD_HOME "/Users/Gnative/Library/Application Support/Herd";
        fastcgi_param REDIRECT_STATUS  200;
        fastcgi_param HTTP_PROXY  "";
        fastcgi_buffer_size 512k;
        fastcgi_buffers 16 512k;
        fastcgi_param SCRIPT_FILENAME "/Applications/Herd.app/Contents/Resources/valet/server.php";
        fastcgi_param PATH_INFO $fastcgi_path_info;
    }

    location ~ /\.ht {
        deny all;
    }
}

```

**Reply by Gnative** [reply]
**Note:** As explained in the Valet documentation this can open your machine up to security risks.

**Reply by Gnative** [reply]
@mpociot would be cool if this was a setting with in Herd Settings to allow opening up of local network.

**Reply by mykkode** [reply]
On windows: `%userprofile%\.config\herd\config\nginx`. Open herd.conf. 

Edit:
    `listen 127.0.0.1:80 default_server;`
to:  
`listen 0.0.0.0:80 default_server;`

You should either create a rule in the firewall or disable it. The best way is to create a rule to allow only the required ports.

**Reply by slavarazum** [reply]
Were you able to successfully run a development build worker with Hot Module Replacement through the local network?

**Reply by Gnative** [reply]
Yes no problem.

**Reply by slavarazum** [reply]
I opened Herd's Nginx server with `listen 80 default_server` and running Vite development build with `--host` option:

```bash
npm run dev -- --host
```

The browser started complaining about the non-private connection, but whatever, we can give the permission.

I'm trying to access with the following url format in Chrome: `http://192.168.*.*/laravel.test`, then browser immediately rewrites it to just `http://192.168.*.*` in address bar for some reason.

In Safari I can't get access for that url at all with the following errors in console:

>[Error] Failed to load resource: bad URL (http://[::]:5173/@vite/client, line 0)
>[Error] Failed to load resource: bad URL (http://[::]:5173/resources/js/app.js, line 0)

Have you encountered similar issues? Perhaps you have any specific Nginx or/and Vite configurations?

**Reply by Gnative** [reply]
Sorry @slavarazum I don't know what your set up is. It does sound like you have some https requests to cause the private connection message and a redirect happening on your app side.. this wouldn't be due to the herd configuration. Wish you luck in figuring this out.

**Reply by ph7jack** [reply]
It looks great. I can load my local website, but the Vite assets (hot or build) never load. Even if changing the APP_URL

**Comment by emandtf** [comment]
Removing "127.0.0.1:" prefix doesn't solves all issues because the URL responds always with "404" if a device's hosts file is not modified by adding a rule for "application_name.test".
I found that this is the culprit file:

- C:\Program Files\Herd\resources\app.asar.unpacked\resources\valet\server.php:45

There are three $SERVER variables that from the development machine have different values than from other devices and when the check for HTTP_HOST is done an URL like "_http://192.168.1.15/application_name.test_" is not verified successfully because the "_$siteName_" is "_192.168.1.15_" and not "_application_name.test_" and then Valet refuses to procede because it doesn't find the right folder.

You can fix this case with one of those cases:

- editing that ("server.php" around row 45) file according
- add a "hosts" rules for you server domain
- add a router Hostname (I'm using OpenWRT): in this way all devices in the same network/LAN/WiFi will found the right path

**Comment by arce701** [comment]
I am writing to request the implementation of a fundamental feature: the ability to share websites within a local network (LAN). This functionality would be decisive in my decision to upgrade to the Pro version of the software.
As a developer working primarily with web applications in corporate environments, we frequently need to test and access applications that operate exclusively on our local network. This need is particularly critical in our location, where the Internet connection is significantly slow and unstable, making work through the Internet impractical and unreliable. The local network allows us to maintain optimal and consistent work speed, independent of external connectivity limitations.
Currently, these limitations force us to continue using alternatives like XAMPP. Implementing this feature in the Pro version would:

1. Facilitate development and testing in business environments.
2. Enhance security by not requiring external exposure.
3. Provide an integrated solution for development teams.
4. Optimize performance in areas with limited connectivity.
5. Enable continuous work independent of Internet quality.

I am convinced that this functionality would not only benefit my team but also many other professionals working under similar conditions, especially in regions where Internet infrastructure is not optimal.

**Reply by emandtf** [reply]
@arce701 I'm developing WITHOUT any internet connection and it already works inside my LAN 😅
I've a test website on my PC and all smartphones (I'm primarily writing mobile apps in native code) in my LAN can reach it and do REST requests.
On your local LAN the "test" website could already been reached by any client.

**Reply by arce701** [reply]
Since you have already successfully implemented it, would you be able to create a tutorial or video sharing your experience and knowledge? It would be extremely helpful to learn the process step by step. In advance, I greatly appreciate your willingness and the time you will dedicate to creating this material...

**Comment by emandtf** [comment]
The solution:

> listen 80 default_server;

is not longer working.
I passed from Herd 1.6 to 1.13.

If I set this config the AdminerEvo login page is displayed.

**Reply by gixxy22** [reply]
did you manage to find any solution to this? having the same issue

**Reply by tompie** [reply]
It's working for me. Modified the `/Users/<username>/Library/Application Support/Herd/config/nginx/herd.conf` to 

`listen 80 default_server;`

Restarted all the services via the Herd menu.

Started `npm run dev --host`

Then on the other client (MacOS) edit `/etc/hosts` and add.

`192.168.1.x yourapp.test` (use your IP and application name)

Not the best solution but works for my requirements. Running Herd 1.14.1 (Build 39).

**Reply by emandtf** [reply]
There isn't any /Library in my User's Home.

The right path would be "_C:\Users\USER\config\herd\config_" where I commented one line and added the second one.

> #listen 127.0.0.1:80 default_server;
listen 80 default_server;

Same procedure for the specific Site config at "_C:\Users\USER\config\herd\config\valet\Nginx_"

> #listen 127.0.0.1:80;
    listen 80;

But nothing helped.
After set those changes at the same "_DOMAIN.test_" address Laravel stopped to respons and now AdminerEvo is responding to it like I show in this picture:
![20250201_herd_localhost_listening](https://github.com/user-attachments/assets/412fc7e7-c993-4737-8965-51ec2fd456a3)


What do you mean with "_Started npm run dev --host_"?? There isn't any way to manually execute it while all Services are already well started.....

**Comment by sakaye** [comment]
I made this work by adding the following lines to my `config/nginx/herd.conf`. I did it this way because I'm testing API requests from another server on my network and wanted it to resolve through https.

```
# New external server blocks
server {
    listen 80 default_server;
    server_name your.server.name;  # Change to your desired external name or IP

    location / {
        proxy_pass http://127.0.0.1:80;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
}

server {
    listen 443 ssl default_server;
    server_name your.server.name;  # Change as needed

    ssl_certificate "/location/to/your/certificate.cer";
    ssl_certificate_key "/location/to/your/certificate_key.key";

    location / {
        proxy_pass https://127.0.0.1:443;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_ssl_verify off;
    }
}
```

On my external server I added an entry to host file to point to the IP address of my local webserver and used the name DNS name I used locally. Example:
`192.168.xx.xx  your-folder-name.test`

Once all that was set up I could visit my local site from another computer on my network and resolve API requests using Postman.

**Reply by emandtf** [reply]
Original file is completly different and doesn't supports SSL:443.......
Why saying "_# New external server blocks_" if you complety rewritten the whole file from scratch?

However this seems not work in any case expecially if no any SSL is used.
This maybe because you're setting NGINX listening to 0.0.0.0:80 and then do a proxy_pass to THE SAME server and :80 port.
It doesn't have any sense for me....
I understand the part fo proxy_pass to :443 that redirects on the other block of Server, but the first block it's complety wrong IMHO.
Proxy_pass is used to pass the request to a **different** combination of IP and/or PORT, not to the same server becuase it hasn't any sense.

PS: even it it could work you would have forced the WHOLE Herd (so all its Sites) to answers to that domain and port.......so you would use/run just one server at a time....

**Reply by sakaye** [reply]
My intention was to open my local dev machine to other computers on my local network. This worked for me so I thought I'd share...

**Comment by pr0ph3tx** [comment]
Unbelievable how this has been open since 2023 and no feature has been developed... Anyway...

Until Herd supports LAN binding natively, here’s a simple workaround to expose specific projects on specific ports (e.g., http://192.168.1.70:5001 → project-x.test, :5002 → project-y.test). This is Windows-focused, but the idea is the same elsewhere.

1. Open Herd’s Nginx folder (Path is usually %USERPROFILE%\.config\herd\config\nginx)
3. Create a file (e.g. mysites.conf) make sure it’s **.conf**, not .conf.txt
4. Add one server block per site/port (duplicate & edit as needed)

```
server {
    listen 0.0.0.0:5001;  #<- YOUR CUSTOM PORT HERE
    server_name _;

    # Send the request to Herd's HTTP on localhost,
    # but tell upstream we want project-x.test
    location / {
        proxy_set_header Host yousite.test;        # <- YOUR HERD SITE HERE
        proxy_set_header X-Forwarded-Host $http_host;# keep client-visible host (IP:port)
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_pass http://127.0.0.1:80;
    }

    access_log off;
}
```

Repeat for 5002 → anothersite.test, etc.

5. Open %USERPROFILE%\.config\herd\config\nginx\nginx.conf
6. Scroll all the way to the bottom and right after "include herd.conf" add "include mysites.conf;"
7. Restart all services

Now you can host any site you have at any custom port. (providing you have the port available)
For others to connect to your sites, simply look for your internal IP address, usually 192.168.X.X and then append your custom port.

Happy development!

**Reply by emandtf** [reply]
I "solved" paying IntelIDEA PhpStorm and just clicking on its "play" button which spawn an HTTP server where provide the whole Laravel site 😅🤷‍♂️

**Comment by KillTomer** [comment]
Solution:
Windows HERD (http, not https) - RUN ALSO ON MOBILE

.env APP_URL=http://192.168.1.x:8000 (PC IP)

vite.config.js
```
server: {
        host: '0.0.0.0',
        hmr: {
            host: '192.168.1.x',
        },
    },
```

Terminal commands:
php -S 192.168.1.x:8000 -t public
npm run dev
